# 🐍 Introduction to Python

## What is Python?

Python is a **high-level**, **interpreted**, and **general-purpose** programming language known for its simplicity and readability. It was created by **Guido van Rossum** and released in **1991**.

Python is great for:
- Web development (using frameworks like Django, Flask)
- Data analysis and visualization
- Artificial Intelligence and Machine Learning
- Scripting and automation
- Game development, IoT, and more!

---

## Why Python?

Before we dive in, let’s understand why Python is such a powerful tool:

- Easy to learn and read 🧠  
- General-purpose and flexible 🧰  
- Widely used in Data Science, AI, Web Dev, Automation 🌐  
- Has an amazing community and tons of free libraries 💾  

Let’s begin our journey by setting up Python and checking our first code!

---

## 🔌 What Are Modules in Python?

A **module** is a file containing Python code — functions, variables, and classes—that you can reuse in other programs.

Python modules help you:
- Keep your code organized
- Avoid writing duplicate code
- Make your projects more modular and readable

You can import modules using the `import` keyword.

```python
import math
```

---

## 🧰 Types of Modules in Python

### 1. **Built-in Modules**
These come with Python out of the box—no installation needed!

Examples:
- `math`: Mathematical functions
- `datetime`: Date and time manipulation
- `random`: Generate random numbers
- `os`: Interact with the operating system

In [ ]:
import datetime
print(datetime.datetime.now())

: 

### 2. **External Modules**
These are not included in Python by default. You install them using **pip**.

Examples:
- `numpy`: Numerical computing
- `pandas`: Data analysis
- `requests`: HTTP requests

```bash
pip install requests
```

In [5]:
import requests
response = requests.get("https://example.com")
response.status_code

200


### 3. **User-Defined Modules**
You can write your own module by saving functions and variables in a `.py` file and importing it into other scripts.

**Example:**
File: `mymodule.py`
```python
def greet(name):
    return f"Hello, {name}!"
```

In another file:
```python
import mymodule
print(mymodule.greet("Alice"))
```

---

## 📦 What is pip?

`pip` stands for **"Pip Installs Packages"**.

It's the **Python package manager** used to install and manage external modules (also called packages or libraries).

### 🔧 Common pip Commands:
```bash
pip install package-name        # Install a package
pip uninstall package-name      # Remove a package
pip list                        # Show installed packages
pip show package-name           # Info about a package
```

> Pro tip: Use `pip freeze > requirements.txt` to save installed packages for project sharing!


> Pro Tip: Use `pip install -r requirements.txt` to install all the packages listed in the `requirements.txt` file.


In [8]:
# Your first spell in Python
print("Hello, Python World!")

Hello, Python World!


## Chapter 1: Understanding the Language of Python

Like any good language, Python has its own alphabet, symbols, and words with special meanings.

### 1. Character Set:
- Letters: a-z, A-Z
- Digits: 0-9
- Special Characters: + - * / = ! @ etc.

### 2. Tokens:
These are the smallest units in a Python program:
- **Keywords**: reserved words (like `if`, `while`, `def`)
- **Identifiers**: names you give to variables and functions
- **Literals**: constant values like numbers or text
- **Operators**: like `+`, `-`, `*`, `/`


# Difference Between Keywords, Identifiers, and Literals

## 🔑 Keywords
- **Definition**: Reserved words in a programming language.
- **Purpose**: Have special meaning and are used to perform specific operations or define structures.
- **Examples (in Python)**:
  - `if`, `else`, `while`, `for`, `def`, `class`, `return`, `import`

> ❌ You cannot use keywords as names for variables or functions.

---

## 🆔 Identifiers
- **Definition**: Names used to identify variables, functions, classes, modules, etc.
- **Rules**:
  - Can include letters (A-Z, a-z), digits (0-9), and underscores (_)
  - Cannot start with a digit
  - Cannot be a keyword
- **Examples**:
  - `myVar`, `calculate_sum`, `UserName`, `_temp`

---

## 🔢 Literals
- **Definition**: Constant values assigned to variables or used directly in code.
- **Types**:
  - **Numeric literals**: `10`, `3.14`, `0xFF`
  - **String literals**: `'Hello'`, `"World"`
  - **Boolean literals**: `True`, `False`
  - **None literal**: `None`

**Example:**

In [ ]:
# Using keywords, identifiers, literals
customer_count = 50  # identifier = literal
if customer_count > 30:  # 'if' is a keyword
    print("Sales are doing well today!")  # literal = string
else:
    print("Need to work on promotions!")

## Identifiers: Naming Your Tools

To name things in Python, follow these rules:
- Can use letters, digits, underscore (`_`)
- Cannot start with a digit
- Cannot use keywords
- Case-sensitive

### Examples:
- ✅ `store_name`, `sales2023`, `_helper`
- ❌ `2store`, `class`, `@value`


In [ ]:
# Examples of valid identifiers
store_name = "Jack's Fabrics"
sales_2023 = 1240


In [18]:
# Invalid ones
# 2store = "Nope"
# class = "Reserved Word"

## Literals & Escape Characters

A **literal** is a fixed value. Python supports:
- **String**: `"Hello"`, `'World'`
- **Number**: `100`, `3.14`
- **Boolean**: `True`, `False`

**Escape Characters** help include special characters in strings:
- `\n` → New line
- `\t` → Tab
- `\\` → Backslash
- `\"` → Quote inside string


In [20]:
print("Welcome to Jack's Store!\nToday's offer:\tBuy 1 Get 1 Free!")
print("He said, \"Python is awesome!\"")

Welcome to Jack's Store!
Today's offer:	Buy 1 Get 1 Free!
He said, "Python is awesome!"


## Comments: Whispering to Yourself

Comments are used to explain your code.

- Single-line: starts with `#`
- Inline: `print("Hi")  # this prints text`
- Multiline: triple quotes `''' comment '''` or `""" comment """`



In [22]:
# This is a single-line comment
print("Comment demo")  # Inline comment

"""
This is a multiline comment
spanning several lines
"""

Comment demo


'\nThis is a multiline comment\nspanning several lines\n'

---

## 🤔 Why Does the Triple-Quoted Text Appear in the Output?

- In **Python**, triple quotes (`'''` or `"""`) are used for:
  1. ✅ **Docstrings** (to describe functions, classes, or modules)
  2. ⚠️ **Multi-line string literals** (which is how they're treated in your code)

- Since the triple-quoted string is not being used (e.g., not assigned to a variable or attached to a function), it's just a **standalone string literal**.

- **Jupyter Notebooks** automatically display the **result of the last expression** — which, in this case, is that string.

---

## ✅ How to Properly Add Multi-line Comments

### Option 1: Use `#` on each line



In [ ]:
# This is a multi-line comment
# that spans multiple lines
# and does not show in output


### Option 2: Suppress triple-quote output by assigning it


In [24]:
_ = """
This is a 'comment' that won't be shown
because it's assigned to a dummy variable
"""

## 🧠 **Chapter 2: Understanding Data Types and Variables in Python**

### 🔤 **What Are Data Types?**

Just like we can describe real-world things (like a name, a number, or a yes/no answer), Python has different types to describe values.

# 🐍 Python Data Types

Python data types are broadly categorized into two types:

- **Primitive Data Types**
- **Non-Primitive (or Complex) Data Types**

---

## 🔹 Primitive Data Types

Primitive data types are the most basic types provided by Python. These are the building blocks for data manipulation.

### 1. **Integers (`int`)**
- Whole numbers (positive or negative)
- Example: `10`, `-5`, `0`

```python
a = 10
```

### 2. **Floating Point Numbers (`float`)**
- Decimal numbers
- Example: `3.14`, `-0.001`, `2.0`

```python
pi = 3.14
```

### 3. **Boolean (`bool`)**
- Represents truth values: `True` or `False`

```python
is_valid = True
```

### 4. **String (`str`)**
- Sequence of characters
- Can be enclosed in single, double, or triple quotes

```python
name = "Alice"
```

---

## 🔸 Non-Primitive Data Types

Non-primitive types are more complex structures that are built using primitive types.

### 1. **List**
- Ordered, mutable collection
- Can contain mixed data types

```python
my_list = [1, 2, "hello", True]
```

### 2. **Tuple**
- Ordered, immutable collection

```python
my_tuple = (1, 2, 3)
```

### 3. **Set**
- Unordered collection of unique elements

```python
my_set = {1, 2, 3, 3}  # Output: {1, 2, 3}
```

### 4. **Dictionary (`dict`)**
- Unordered collection of key-value pairs

```python
my_dict = {"name": "Bob", "age": 25}
```

## 🔄 What Does **Mutable** Mean?

- **Mutable** means something that **can be changed or modified after it's created**.
- **Immutable** means it **cannot be changed** after it's created.

Think of it like:

| Type      | Can it change? | Example                     |
|-----------|----------------|-----------------------------|
| Mutable   | ✅ Yes          | List (`[1, 2, 3]`)           |
| Immutable | ❌ No           | Integer (`5`), String (`"hi"`) |

---

## 🔁 Summary Table

| Category       | Type     | Mutable | Ordered | Example              |
|----------------|----------|---------|---------|----------------------|
| Primitive       | `int`    | ❌      | ✅      | `10`                 |
| Primitive       | `float`  | ❌      | ✅      | `3.14`               |
| Primitive       | `bool`   | ❌      | ✅      | `True`               |
| Primitive       | `str`    | ❌      | ✅      | `"Hello"`            |
| Non-Primitive   | `list`   | ✅      | ✅      | `[1, 2, 3]`          |
| Non-Primitive   | `tuple`  | ❌      | ✅      | `(1, 2, 3)`          |
| Non-Primitive   | `set`    | ✅      | ❌      | `{1, 2, 3}`          |
| Non-Primitive   | `dict`   | ✅      | ❌ (as of <3.7) | `{"a": 1}`     |

## 🧠 Why Are **Primitive Data Types Immutable** in Python?

### 1. **Simplicity and Safety**
Immutable objects are safer to use because they **can’t be accidentally modified**, especially when passed around in your code.

```python
a = 10
b = a
b = 20
print(a)  # Still 10
```

> Changing `b` didn’t affect `a`, because `int` is immutable.

---

### 2. **Memory Efficiency (Interning)**
Python **reuses** some immutable objects (like small integers and strings) to save memory.

```python
a = 5
b = 5
print(a is b)  # True — same memory reference
```

If they were mutable, this optimization would be risky because one change would affect all references.

---

### 3. **Hashability (Important for Dict Keys & Set Elements)**
Immutable objects can be used as keys in dictionaries or elements in sets.

```python
my_dict = {"name": "Alice"}  # 'name' is a str (immutable)
```

Mutable objects can’t be used this way because their value might change, breaking the dictionary's structure.

---

## 🧪 Real-World Analogy

Imagine:

- **Immutable**: A printed receipt — you can't change the values after it's printed.
- **Mutable**: A whiteboard — you can erase and change it anytime.

---

Python makes primitive types immutable for:
- Stability
- Performance
- Safe memory handling


In [30]:
# Here's Alex exploring basic data types
x = 10             
price = 19.99      
name = "Alex"     
is_active = True

In [32]:
print(type(x))
print(type(price))
print(type(name))
print(type(is_active))

<class 'int'>
<class 'float'>
<class 'str'>
<class 'bool'>


### 📦 **Variables in Python**

I just like how simple python makes variable creation.

🧑‍🏫 *In languages like Java or C++, you'd have to define types before assigning them. Not in Python!*


In [ ]:
# Just assign a value and Python figures out the type
score = 85
greeting = "Hello"
print(type(score))
print(type(greeting))

Python manages memory dynamically and even allows reassigning values of a different type to the same variable!

In [ ]:
score = "Eighty Five"
print(type(score))

### 🔢 **Numbers: Integers and Floats**


In [34]:
a = 100        # Integer
b = -25        # Negative Integer
c = 3.14159    # Float

In [36]:
rounded_pi = round(c, 2)  # 3.14
print(rounded_pi)

3.14


# 🧮 Integer Division vs Float Division


In [38]:
print(10 / 5)

2.0


In [40]:
print(10 // 3) #(floor division)

3


# ✍️ Strings: More Than Just Words

In [42]:
word1 = 'Panem'
word2 = "Canal"
sentence = "Panama Canal"

In [44]:
name = word1 + " " + word2
name

'Panem Canal'

In [46]:
len(name)

11

# 🎯 Indexing and Slicing

In [48]:
name[0]

'P'

In [ ]:
name[0:6]

In [50]:
# ❓ Booleans
is_numeric = True
print(is_numeric, type(is_numeric))

True <class 'bool'>


In [52]:
greater_than_50 = 75 > 50
print(greater_than_50)

True


In [54]:
# 🔄 Type Conversion
x = "100"
y = float(x)
y

100.0

In [56]:
z = int(float("7.9"))
z

7

In [60]:
# Watch out:
# int("abc")

# 🌫️ What’s None?

In [62]:
temp = None
print(temp, type(temp))

None <class 'NoneType'>


In [64]:
from math import isnan
import math

missing_value = float('nan')
print(missing_value)
print(isnan(missing_value))

nan
True


## 🔹 `NoneType`
- **`None`** is the **only value** of the **`NoneType`**.
- Represents **absence of a value** or **"nothing"**.
- Often used as a **default return** when no value is explicitly returned from a function.


### Common Uses:
- Optional function return
- Placeholder for missing data
- Default argument value in functions

---

## 🔹 `NaN` (Not a Number)
- Stands for **"Not a Number"**
- A special **floating-point value** used to represent:
  - Missing or undefined numerical data
  - Invalid mathematical results (like `0/0`)
- Comes from `float('nan')` or some data libraries like NumPy or pandas.

### Key Properties:
- Type: `<class 'float'>`
- `NaN != NaN` (strange but true!)
- Use `math.isnan()` or `numpy.isnan()` to check for it

---

# 🎤 Getting User Input


In [ ]:
name = input("What is your name? ")
age = int(input("What is your age? "))
print(f"{name} is {age} years old.")

# 🖨️ Printing and Formatting Output

In [66]:
name = "Alex"
age = 23

print("Hello", name, "you are", age, "years old.")  # Using commas
print("Hello {} you are {} years old.".format(name, age))  # Using format()
print(f"Hello {name}, you are {age} years old.")  # Using f-string (Python 3.6+)

Hello Alex you are 23 years old.
Hello Alex you are 23 years old.
Hello Alex, you are 23 years old.


## Separator and end:

In [68]:
print("Python", "is", "fun", sep="$", end="!\n")

Python$is$fun!


# 🧮 Operators in Python

In [70]:
x, y = 5, 3
print(x + y, x - y, x * y, x / y, x // y, x % y, x ** y,sep="\n", end="\n\n")
print(x > y, x < y, x == y, x != y,sep="\n", end="\n\n")
print(x > 0 and y > 0)
print(x is y, x is not y,sep="\n", end="\n\n")
print('a' in 'apple', 'z' not in 'apple',sep="\n", end="\n\n")

8
2
15
1.6666666666666667
1
2
125

True
False
False
True

True
False
True

True
True



## ➕ Arithmetic Operators
| Expression   | Result | Meaning                |
|--------------|--------|------------------------|
| `x + y`      | `8`    | Addition               |
| `x - y`      | `2`    | Subtraction            |
| `x * y`      | `15`   | Multiplication         |
| `x / y`      | `1.67` | Division (float)       |
| `x // y`     | `1`    | Floor division (int)   |
| `x % y`      | `2`    | Modulo (remainder)     |
| `x ** y`     | `125`  | Exponentiation (power) |

---

## ⚖️ Comparison Operators
| Expression    | Result | Meaning                |
|---------------|--------|------------------------|
| `x > y`       | `True` | Greater than           |
| `x < y`       | `False`| Less than              |
| `x == y`      | `False`| Equal to               |
| `x != y`      | `True` | Not equal to           |

---

## 🔗 Logical Operators
| Expression              | Result | Meaning                     |
|--------------------------|--------|-----------------------------|
| `x > 0 and y > 0`        | `True` | Both conditions are `True` |

---

## 🆔 Identity Operators
| Expression         | Result | Meaning                      |
|--------------------|--------|------------------------------|
| `x is y`           | `False`| Checks if `x` and `y` are the same object |
| `x is not y`       | `True` | They are different objects   |

---

## 🔤 Membership Operators
| Expression                  | Result | Meaning                   |
|-----------------------------|--------|---------------------------|
| `'a' in 'apple'`            | `True` | `'a'` is in the string     |
| `'z' not in 'apple'`        | `True` | `'z'` is **not** in the string |

---

## 🔲 Bitwise Operators
| Expression          | Result  | Meaning                       |
|---------------------|---------|-------------------------------|
| `x & y`            | `1`     | Bitwise AND                   |
| `x \| y`            | `7`     | Bitwise OR                    |
| `x ^ y`            | `6`     | Bitwise XOR                   |
| `~x`               | `-6`    | Bitwise NOT (complement)      |
| `x << 1`           | `10`    | Left shift                    |
| `x >> 1`           | `2`     | Right shift                   |

---

## 📝 Assignment Operators
| Expression          | Result | Meaning                        |
|---------------------|--------|--------------------------------|
| `x = y`             | `x` = 3 | Assign `y` to `x`              |
| `x += y`            | `x` = 8 | `x = x + y`                    |
| `x -= y`            | `x` = 2 | `x = x - y`                    |
| `x *= y`            | `x` = 15| `x = x * y`                    |
| `x /= y`            | `x` = 1.67 | `x = x / y`               |
| `x //= y`           | `x` = 1 | `x = x // y`                  |
| `x %= y`            | `x` = 2 | `x = x % y`                   |
| `x **= y`           | `x` = 125| `x = x ** y`                  |
| `x &= y`            | `x` = 1 | `x = x & y`                   |
| `x \|= y`            | `x` = 7 | `x = x \| y`                   |
| `x ^= y`            | `x` = 6 | `x = x ^ y`                   |
| `x <<= 1`           | `x` = 10| `x = x << 1`                  |
| `x >>= 1`           | `x` = 2 | `x = x >> 1`                  |

---

## 📊 Summary Table of All Operators

| Type               | Operators                                                  |
|--------------------|------------------------------------------------------------|
| **Arithmetic**      | `+`, `-`, `*`, `/`, `//`, `%`, `**`                        |
| **Comparison**      | `>`, `<`, `==`, `!=`                                       |
| **Logical**         | `and`, `or`, `not`                                         |
| **Identity**        | `is`, `is not`                                             |
| **Membership**      | `in`, `not in`                                             |
| **Bitwise**         | `&`, `^`, `~`, `<<`, `>>` ,`\|`                            |
| **Assignment**      | `=`, `+=`, `-=`, `*=`, `/=`, `//=`, `%=` , `**=`, `&=`, `\|=`, `^=`, `<<=`, `>>=` |


## 🧠 Story: A Day in Everyone's Life – Decisions, Decisions...

Every morning, even before we leave the house, we've already made a dozen decisions.  
Tea or coffee? Shave or not? Blue shirt or beige? Sweater or no sweater?

All these choices can be boiled down to simple **true or false** questions. That’s exactly how computers think too.


## 🔍 Part 1: Conditional Statements – Making Decisions


In [73]:
weather_is_cold = True

if weather_is_cold:
    print("Wear a sweater")
else:
    print("No need for a sweater")


Wear a sweater


🧵 **Explanation**:
- `if` checks a condition.
- If `True`, it runs the indented block under `if`.
- If `False`, it skips to `else`.


## 🧪 Let’s simulate more decisions:

I decide whether to have cook breakfast or have outside based on how much time i have.

In [75]:
time_for_breakfast = 10  # Time i have before i leave

if time_for_breakfast >= 15:
    print("Cook myself")
else:
    print("Grab on my way")

Grab on my way


## 🔁 Part 2: Loops – Repeating Actions

Imagine John’s fitness coach tells him:

> Walk around the park **8 times**.

This is a **definite loop** – we know the number of repetitions.


In [ ]:
for round in range(1, 9):
    print(f"Round {round}: Walking...")
print("Done with all 8 rounds!")



Now imagine the coach says:

> Walk **until** you reach the ice cream truck.

That’s an **indefinite loop** – we don’t know how long it’ll take.



In [ ]:
truck_spotted = False
steps = 0

while not truck_spotted:
    steps += 1
    print(f"Step {steps}: Still walking...")
    if steps == 7:
        truck_spotted = True

print("Ice cream truck found! 🍦")


# 🔁 Python `for` Loop: Different Ways to Use It

The `for` loop in Python is super versatile and lets you iterate over sequences like lists, strings, tuples, dictionaries, and more.

---

## ✅ Basic Loop Over a List

In [80]:
fruits = ["apple", "banana", "cherry"]
for fruit in fruits:
    print(fruit)

apple
banana
cherry


## 🔢 Loop with `range()`

In [85]:
for i in range(1, 10, 2):
    print(i)

1
3
5
7
9


| Method      | Resulting Range                  |
|:------------------|:------------------------------|
| `range(n)`      | 0 to n-1                    |
| `range(2, 6)`    | 2 to 5                       |
| `range(1, 10, 2)`| 1 to 9 with step of 2        |

---

## 🔁 Loop Over a String

In [87]:
for letter in "hello":
    print(letter)

h
e
l
l
o


## 🎯 Loop with `enumerate()`: Get Index and Value

In [90]:
colors = ["red", "green", "blue"]
for index, color in enumerate(colors):
    print(index, color)

0 red
1 green
2 blue


## 🔑 Loop Over a Dictionary

In [92]:
person = {"name": "Alice", "age": 25}
for key in person:
    print(key, person[key])

name Alice
age 25


In [94]:
# Better: use .items()
for key, value in person.items():
    print(key, value)

name Alice
age 25


## 🔄 Nested `for` Loop

In [96]:
for i in range(3):
    for j in range(2):
        print(i, j)

0 0
0 1
1 0
1 1
2 0
2 1


## 📦 Loop Over a Tuple

In [98]:
coords = (1, 2, 3)
for c in coords:
    print(c)

1
2
3


## 🧹 Loop Over a Set (Unordered, Unique Items)

In [100]:
items = {"a", "b", "c"}
for item in items:
    print(item)

b
a
c


## 🗂️ Loop with List Comprehension (One-liner)

In [105]:
squares = [x**2 for x in range(5)]
print(squares)

[0, 1, 4, 9, 16]


## 🔄 `for` with `else`

In [107]:
for i in range(5):
    print(i)
else:
    print("Loop completed!")

0
1
2
3
4
Loop completed!


> `else` runs **only if the loop wasn't broken by `break`**.

In [113]:
for i in range(5):
    print(f"Checking {i}")
    if i == 3:
        print("Breaking the loop!")
        break
        #continue
else:
    print("Loop completed without break.")

Checking 0
Checking 1
Checking 2
Checking 3
Breaking the loop!


## ⛔ With `break` and `continue`

In [115]:
for i in range(5):
    if i == 3:
        break   # stops loop
    print(i)

0
1
2


In [117]:
for i in range(5):
    if i == 2:
        continue  # skips 2
    print(i)

0
1
3
4


## 🧠 Summary

| Use Case                    | Example                            |
|-----------------------------|-------------------------------------|
| List                       | `for x in [1,2,3]:`                |
| String                     | `for c in "abc":`                  |
| Dictionary                 | `for k, v in dict.items():`        |
| `range()`                  | `for i in range(5):`               |
| Index + Value              | `for i, val in enumerate(lst):`    |
| Nested Loop                | `for x in a: for y in b:`          |
| List Comprehension         | `[x for x in range(5)]`            |
| `for...else`               | `for x in lst: ... else:`          |
| With `break`, `continue`   | Custom control of loop flow        |


# 🔁 Converting a `for` Loop to a `while` Loop in Python

Python's `for` loops are great for iterating over sequences or ranges. But any `for` loop can be rewritten using a `while` loop — you just need to manage the **loop variable**, **condition**, and **updates** manually.

---

## 🧪 Example 1: Using `range()`

### ✅ `for` version:

In [119]:

for i in range(5):
    print(i)


0
1
2
3
4


### 🔁 `while` version:

In [121]:
i = 0
while i < 5:
    print(i)
    i += 1

0
1
2
3
4


## 🧪 Example 2: Looping Over a List

### ✅ `for` version:

In [123]:
fruits = ["apple", "banana", "cherry"]
for fruit in fruits:
    print(fruit)

apple
banana
cherry


### 🔁 `while` version:

In [125]:
fruits = ["apple", "banana", "cherry"]
i = 0
while i < len(fruits):
    print(fruits[i])
    i += 1

apple
banana
cherry


## 🧪 Example 3: `for` with `break`

### ✅ `for` version:

In [127]:
for i in range(5):
    if i == 3:
        break
    print(i)

0
1
2


### 🔁 `while` version:

In [129]:
i = 0
while i < 5:
    if i == 3:
        break
    print(i)
    i += 1

0
1
2


## 💡 Tips for Converting `for` ➡️ `while`

| Aspect             | `for` Loop                      | `while` Loop                        |
|--------------------|----------------------------------|--------------------------------------|
| Initialization     | Implicit (inside `range()`)     | You must do it manually              |
| Condition          | Built into `range()` or iterable| Must write explicitly (`i < n`)      |
| Update             | Implicit in `for`               | Must do manually (`i += 1`)          |
| Indexing (lists)   | Implicit in `for item in list`  | Use `list[i]` with manual counter    |


# 🔁 `do...while` in Python? Let's Talk About It

Python **does not have** a built-in `do...while` loop like some other languages (e.g., C, C++, JavaScript). But the idea can still be mimicked using a `while` loop.

## ❓ What is `do...while`?

In many languages, `do...while` runs the **loop body first**, and **then checks the condition**.

### Example in C-style syntax:
```c
do {
    // code runs at least once
} while (condition);
```

---

## 🔁 Mimicking `do...while` in Python

You can simulate this behavior by using a `while` loop with a condition that is **initially True**, then breaking out when needed.

### ✅ Python Example:

In [ ]:
while True:
    user_input = input("Enter something (q to quit): ")
    if user_input == 'q':
        break
    print("You typed:", user_input)

🔹 This runs the loop **at least once**, just like a `do...while`.

---

## 🤷 Why Python Doesn’t Have `do...while`

- Python favors **clarity and simplicity**.
- You can achieve the same behavior with a regular `while True` + `break`.
- Adding a special loop type for a rare case would go against Python's "one obvious way to do it" philosophy (Zen of Python).

---

## 🧠 When to Use This Pattern

- When the loop must **run at least once** (e.g., user input, menu systems)
- When using `while` feels a bit awkward but still works



| Feature             | `do...while`         | Python Equivalent                  |
|---------------------|----------------------|-------------------------------------|
| Body runs first     | ✅                   | ✅ with `while True`                |
| Condition after body| ✅                   | ✅ using `break` inside             |
| Built-in Syntax     | ✅ in other languages| ❌ (no `do...while` in Python)      |



So while Python doesn’t *officially* have `do...while`, you can still use the same logic easily! 🔁🐍

## 🧠 Key Concepts Recap

| Concept          | Description |
|------------------|-------------|
| `if`, `elif`, `else` | For decision-making based on conditions |
| `for` loop        | Use when you know how many times to repeat |
| `while` loop      | Use when repetition depends on a condition |


## 🔸 1. **Comparison Operators**  
(*==, !=, <, >, <=, >=*)

### 🧠 Concept:
These are used to compare values and return `True` or `False`.

### 🎬 Story: *John’s Gym Entry Check*

> John arrives at the gym and checks his body temperature. If it’s **equal to** 98.6°F, he’s good to go. If it’s **higher than 100°F**, he decides to skip the workout. If it’s **less than 98°F**, he puts on a hoodie and proceeds cautiously.


In [135]:
temperature = float(input("Enter your body temperature in Fahrenheit: "))

if temperature == 98.6:
    print("You're in perfect shape! Hit the gym!")
elif temperature > 100:
    print("You might have a fever. Skip the workout today.")
elif temperature < 98:
    print("It’s a bit low. Put on a hoodie and proceed with caution.")
else:
    print("Go to gym")
    


Enter your body temperature in Fahrenheit:  100


Go to gym


```
  |-- |----         |        ------|
<98  98.6        <100>             100+  
```

- `==` to compare equality
- `>` and `<` for thresholds
- Chain of conditions using `elif`

---

## 🔸 2. **Logical Operators**  
(*and, or, not*)

### 🧠 Concept:
Used to combine multiple Boolean conditions.

### 🎬 Story: *John’s Weather + Mood Check*

> Before heading out, John asks:
> - Is it cold **and** raining? He’ll take a cab.
> - Is it cold **or** he’s feeling lazy? He’ll skip the park walk.
> - If it's **not** raining, he might consider walking.



In [140]:
is_cold = input("Is it cold? (yes/no): ") == "yes" 
is_raining = input("Is it raining? (yes/no): ") == "yes"
is_lazy = input("Are you feeling lazy? (yes/no): ") == "yes"

# print(type(is_cold))
# print(is_cold)

if is_cold and is_raining:
    print("Take a cab today.")
elif is_cold or is_lazy:
    print("Maybe skip the walk today.")
elif not is_raining:
    print("Looks good! Go for a walk.")


Is it cold? (yes/no):  yes
Is it raining? (yes/no):  
Are you feeling lazy? (yes/no):  


<class 'bool'>
True
Maybe skip the walk today.


## 🔸 3. **Nested Conditionals**

### 🧠 Concept:
`if` statements inside other `if` blocks.

### 🎬 Story: *John Deciding Breakfast Based on Diet Plan*

> John is on a low-carb diet.  
> - If he’s following his diet today:  
>   - If it's Sunday, he has a cheat day: eat pancakes.  
>   - Else, just have eggs.  
> - If he’s not dieting, he eats whatever he wants.


In [ ]:
on_diet = input("Are you on a diet today? (yes/no): ") == "yes"
day = input("What day is it today?: ")

if on_diet:
    if day.lower() == "sunday":
        print("Cheat day! Enjoy your pancakes.")
    else:
        print("Stick to eggs today.")
else:
    print("Eat anything you like!")


## 🔸 4. **`break` and `continue` in Loops**

### 🧠 Concept:
- `break` exits the loop early
- `continue` skips current iteration but keeps looping

### 🎬 Story: *John Walking Until He Finds the Ice Cream Truck*

> John takes a walk around the block up to 10 times.  
> - If he hears the ice cream truck, he stops (`break`).  
> - If it's raining during a round, he skips that round (`continue`) and runs under shelter.



In [144]:
for round in range(1, 11):
    print(f"Round {round} started.")
    raining = input("Is it raining? (yes/no): ") == "yes"
    if raining:
        print("It's raining, skipping this round.")
        continue

    truck_heard = input("Do you hear the ice cream truck? (yes/no): ") == "yes"
    if truck_heard:
        print("Yay! Ice cream truck found. Stopping walk.")
        break

    print(f"Finished round {round}.")


Round 1 started.


Is it raining? (yes/no):  yes


It's raining, skipping this round.
Round 2 started.


Is it raining? (yes/no):  yes


It's raining, skipping this round.
Round 3 started.


Is it raining? (yes/no):  no
Do you hear the ice cream truck? (yes/no):  yes


Yay! Ice cream truck found. Stopping walk.


- Using `break` for early exit
- Using `continue` to skip specific logic in a loop


## 🔸 5. **Input Validation**

### 🧠 Concept:
Always check that user input is valid before processing.


In [168]:
# celsius_input = float(input("Enter temperature in Celsius: "))
# print(type(celsius_input))
# celsius_input
celsius_input = input("Enter temperature in Celsius: ")
if celsius_input.replace('.', '', 1).isdigit():
    celsius = float(celsius_input)
    fahrenheit = celsius * 1.8 + 32
    print("Temperature in Fahrenheit: %.1f" % fahrenheit)
else:
    print("Invalid input! Please enter a number.")


Enter temperature in Celsius:  23


Temperature in Fahrenheit: 73.4


### 🧠 Note: Why Not Use `.isfloat()` in Python?

#### 🔹 Common Question:
> "Why not just use `isfloat()` to check if a string is a valid float?"

#### ❌ The Issue:
Python **does not have a built-in** `.isfloat()` method for strings like it does with `.isdigit()`.

So this won't work:
```python
"23.5".isfloat()  # ❌ AttributeError
```

---

### ✅ The Alternative: Use `try/except` with `float()`

Since there's no `.isfloat()` method, the most reliable way to check if a string is a float is by attempting to convert it:

```python
def isfloat(value):
    try:
        float(value)
        return True
    except ValueError:
        return False
```

#### 🔍 Why this works:
- ✔ Recognizes valid floats like `"23.5"`, `"0.0"`, `"-7.1"`
- ✔ Accepts scientific notation like `"1e3"`
- ❌ Rejects invalid strings like `"23.5.1"` or `"abc"`

---

### 🛠️ What's this line doing?

```python
celsius_input.replace('.', '', 1).isdigit()
```

#### It tries to simulate float checking:
- Removes **one dot** (e.g., `"23.5"` → `"235"`)
- Checks if the result is all digits
- ✅ Works for basic cases like `"23.5"`
- ❌ Fails for:
  - Negative numbers (`"-3.5"`)
  - Scientific notation (`"1e3"`)
  - Multiple dots (`"12.3.4"`)

---

### ✅ Best Practice:
If you need to check whether a string is a float — **reliably** and **safely** — define and use an `isfloat()` function with `try/except`.

## 🌿 **Branching Statements in Python**

Welcome to the session on **branching statements**!  
After completing this module, you will be able to:
- Distinguish between various conditional statements
- Choose when to use nested conditional statements
- Express conditional logic using logical operators

### 🧩 What Are Conditional Statements?

Conditional statements form the **building blocks** of any programming language.  
They are used to **test assumptions or compare values** to decide what action should be taken.

Python uses **indentation** to determine the scope of code blocks, just like loops.

---

### ✅ Basic `if` Statement

The basic syntax involves using the `if` keyword with a logical comparison:

In [ ]:
x = 5
if x == 5:
    print("x is 5")

If the condition evaluates to `True`, the indented code block is executed.  
Otherwise, it is skipped.

---

### ❗ `else` Block

To handle the scenario when the condition fails, we can use an `else` block:


In [ ]:
x = 3
if x == 1:
    print("x is 1")
else:
    print("x is not 1")

You can also write an `if` statement **without an `else`**, if there's nothing to do when the condition fails.

---

### 🔁 Nested Conditionals

You can **nest `if` statements** inside one another:

In [170]:
x = 10
if x > 5:
    if x < 15:
        print("x is between 5 and 15")

x is between 5 and 15


It’s not mandatory to pair every `if` with an `else`.

---

### 🌟 Single-Line Conditional (Ternary Operator)

For compact conditions:

In [182]:
x=4
result = "Yes" if x > 5 else "No"
print(type(result))
print(result)

<class 'str'>
No


---

### 🧠 Using `elif` for Multiple Conditions

To check **more than one condition**:

In [ ]:
if x > y:
    print("x is greater")
elif x < y:
    print("x is smaller")
else:
    print("x and y are equal")

You can have **multiple `elif` blocks**, but only **one `else`** at the end.

---

### 🔗 Logical Operators in Conditions

To combine multiple comparisons:

In [ ]:
if (x > 5) and (y > 5):
    print("Both are greater than 5")
elif (x > 5) or (y > 5):
    print("At least one is greater than 5")
else:
    print("Neither is greater than 5")

Use `and`, `or`, and `not` for logical operations.

---

### 🧪 Example

In [ ]:
x = 7.5
y = "Hello"

if isinstance(x, float) and isinstance(y, float):
    print("Compare floats")
elif isinstance(x, str) and isinstance(y, str):
    print("Concatenate strings")
else:
    print("Mismatch in types")

## The Ternary Operator (Conditional Expression)

The ternary operator is a shorthand way to write an `if-else` statement. It is used to return one of two values based on a condition.

### Syntax:
```python
value_if_true if condition else value_if_false
```

### Example:

In [184]:
x = 10
result = "Greater than 5" if x > 5 else "Less than or equal to 5"
print(result)

Greater than 5


### Explanation:
- If the condition `x > 5` is `True`, the value `"Greater than 5"` is assigned to `result`.
- If the condition is `False`, `"Less than or equal to 5"` is assigned to `result`.

### Another Example:

In [186]:
x = 3
result = "Positive" if x > 0 else "Non-positive"
print(result)

Positive


### Explanation:
- If `x > 0` is `True`, `result` will be `"Positive"`, otherwise it will be `"Non-positive"`.

## Multiple Conditions with Ternary Operator

You can also chain ternary operators to handle more than two conditions.

### Syntax:
```python
value_if_true if condition else value_if_false if condition2 else value_if_false2
```

### Example:

In [188]:
x = 5
result = "Greater than 10" if x > 10 else "Greater than 0" if x > 0 else "Less than or equal to 0"
print(result)

Greater than 0


### Explanation:
- The first condition checks if `x > 10`, and if `False`, it checks the second condition (`x > 0`).
- Depending on which condition is `True`, the corresponding value is returned.

## Summary

- **`if` statement**: Executes a block of code if a condition is `True`.
- **`if-else` statement**: Executes one block of code if the condition is `True`, and another if it's `False`.
- **`elif`**: Checks multiple conditions in sequence.
- **Nested `if`**: Allows you to check conditions inside other conditions.
- **Logical operators**: Use `and`, `or`, `not` to combine conditions.
- **Ternary operator**: A shorthand for `if-else` that returns a value based on a condition.
- **Multiple conditions with ternary operator**: Chain ternary operators to evaluate more than one condition.